In [41]:
import numpy as np
from scipy.stats import poisson
from scipy.stats import expon
from scipy.stats import binom
from scipy.stats import norm
from scipy.stats import geom

# You are working as a data engineer who has to resolve any bugs
# failures of ML models in production.
# The time taken to debug is exponentially distributed with mean of 5 minutes
# Find the probability of debugging in 4 to 5 minutes?

In [7]:
# 1 bug -> 5 minutes
# x bugs -> 1 minutes
# x(rate) -> 1/5

rate = 1/5
scale = 1/rate
# scale = 5

# P([4<X<5]) = CDF(5) - CDF(4)
expon.cdf(5,scale=scale) - expon.cdf(4,scale=scale)

np.float64(0.08144952294577923)

# Find the probability of needing more than 6 minutes to debug

In [8]:
1-expon.cdf(6,scale=scale)

np.float64(0.3011942119122022)

# Given that you have already spent 3 minutes, what is the probability of needing more than 9 minutes

In [11]:
# P[T>9|T>3] = P[(T>9) and (T>3)]/P(T>3) = P(T>9)/P(T>3)
(1-expon.cdf(9,scale=scale))/(1-expon.cdf(3,scale=scale))

#  same as P[T>6]
1-expon.cdf(6,scale=scale)

np.float64(0.3011942119122022)

A transportation analyst models flight delays (in minutes) at a major airport. She collects the following data:

70% of flights are on time (0 minutes delay)
15% are delayed by exactly 30 minutes
15% are delayed by exactly 60 minutes
She reports this using a CDF plot to her non-technical team.

Later, another analyst claims:

"We can also derive a PDF and compute the probability that a flight is delayed by more than 30 minutes."

Question:
What is the correct interpretation of this situation?

Options:
A. This is a PDF problem since the time is continuous
B. The CDF allows computation of
         P(X>30)=1−CDF(30)=0.15 
C. This is a PMF, so using CDF doesn’t make sense
D. The correct probability for
         P(X>30)=0.3 


In [14]:
# for discrete CDF holds good
# probability in 30 minutes delayed  P(X<=30)  CDF(30)
# P(X<=30) = P(0)+P(30)
cdf30 = 0.7+0.15
result = 1 - cdf30
result

0.15000000000000002

Question 2:

An insurance company receives 20 claims daily. The probability a claim is fraudulent is 0.15.

Find the probability that:

A. At least 5 claims are fraudulent
B. Exactly 3 claims are fraudulent
C. Expected fraudulent claims and its variance
Options:

A. A = 0.38, B = 0.25, Exp = 3, Var = 2.55
B. A = 0.17, B = 0.24, Exp = 3, Var = 2.55
C. A = 0.46, B = 0.25, Exp = 2, Var = 1.7
D. A = 0.28, B = 0.26, Exp = 3, Var = 2.1

In [23]:
#P(x>=5) = 1 - P(x<=4)
A = 1-binom.cdf(k=4,n=20,p=0.15)

B = binom.pmf(k=3,n=20,p=0.15)

ExpectedFraudClaims = 20*0.15 # sample*probability

variance = 20*0.15*0.85  # n*p*(1-p)

print(f'values of A and B are {A} and {B}')
print(f'Expected number of fraud claims are {ExpectedFraudClaims} and variance is {variance}')

values of A and B are 0.1701531537313461 and 0.24282889614926748
Expected number of fraud claims are 3.0 and variance is 2.55


Question 3:

Scenario:
A YouTuber earns revenue based on the number of ads watched. Each ad generates an amount between ₹0.50 and ₹3 randomly.
The revenue per ad follows a Uniform Distribution:X∼U(0.5,3)
X∼U(0.5,3)
The YouTuber uploads 5 videos, and each video gets an average of 1,000 ad views.
- Question 1: What is the expected revenue per ad?
- Question 2: What is the expected total revenue from all 5 videos?

Options:

A. Revenue per ad = ₹1.00, Total revenue = ₹5,000
B. Revenue per ad = ₹1.75, Total revenue = ₹8,750
C. Revenue per ad = ₹2.50, Total revenue = ₹12,500
D. Revenue per ad = ₹3.00, Total revenue = ₹15,000

In [48]:
# mean = (A+B)/2
a = 0.5
b = 3
mean = (a + b)/2
print(f'expected revenue per add is {mean}')
num_videos = 5
num_views = 1000
expected_views = mean * num_videos * num_views
print(f'Expected number of views are {expected_views}')

expected revenue per add is 1.75
Expected number of views are 8750.0


Scenario: A user plays a Google Pay cashback scratch card where they win ₹10, ₹20, ₹50, ₹100, or ₹500, each with an equal chance.

Question 1: What is the probability of winning ₹50? (Use PMF)

Question 2: What is the expected cashback a user will receive? (Use Expectation)


Options:

A. Probability = 0.25, Expected cashback = ₹136

B. Probability = 0.20, Expected cashback = ₹136

C. Probability = 0.10, Expected cashback = ₹100

D. Probability = 0.20, Expected cashback = ₹50

In [ ]:
prob = 1/5 # 1 of 5 equal events
mean = (10+20+50+100+500)/5
# np.mean([10,20,50,100,500])
print(f'Mean is {mean}')

Mean is 136.0


Scenario: An Uber driver tracks ride durations in a city. Data shows that rides follow a normal distribution with:

Mean ride time (μ) = 15 minutes
Standard deviation (σ) = 3 minutes
Question 1: What is the probability that a randomly selected ride takes more than 20 minutes?

Question 2: What is the expected ride duration, and what is the variance?

Question 3: What is the probability that a ride takes between 12 and 18 minutes?

Options:

A. P(X>20)=0.05, Expected = 15 min, Variance = 9, P(12<X<18)=0.68

B. P(X>20)=0.10, Expected = 18 min, Variance = 3, P(12<X<18)=0.50

C. P(X>20)=0.03, Expected = 15 min, Variance = 6, P(12<X<18)=0.70

D. P(X>20)=0.15, Expected = 12 min, Variance = 9, P(12<X<18)=0.60

In [ ]:
mean = 15
std_dev =3

# P(x>20)
q1 = 1 - norm.cdf(20, loc=mean, scale=std_dev)
print(f'Probability that x>20 is {q1}')

# std_dev =sqrt(var)
q2 = std_dev**2
print(f'Variance is {q2}')

q3 = norm.cdf(18,loc=mean, scale=std_dev) - norm.cdf(12, loc=mean, scale=std_dev)
print(f'Probability that 12<x<18 is {q3}')

# A is the anser

Probability that x>20 is 0.047790352272814696
Variance is 9
Probability that 12<x<18 is 0.6826894921370859


# Suppose you are playing a game where success rate of winning a prize is 0.7.
# (a) What is the probability of winning a prize on 4th attempt?
# (b) What is the probability that you don't win in 2 attempts?
# (c) What is the expected number of trials to get the first success?

In [39]:
a = geom.pmf(k=4, p=0.7)
print(f'Prob of winning on 4th attempt is {a}')

b = 1 - geom.cdf(k=2, p=0.7)
print(f'Prob of not winning on 2 attempts is {b}')

c = 1/0.7
print(f'Expected number of attempts to win is {c}')

Prob of winning on 4th attempt is 0.018900000000000007
Prob of not winning on 2 attempts is 0.08999999999999997
Expected number of attempts to win is 1.4285714285714286


Question 11:

Scenario: A WhatsApp user receives an average of 5 messages per minute. The number of messages follows a Poisson distribution with λ=5.

Question 1: What is the probability that the user receives exactly 7 messages in a minute?

Question 2: What is the probability that the user receives at least 3 messages in a minute?


A. P(X=7) = 0.1044, P(X≥3) = 0.8753

B. P(X=7) = 0.1200, P(X≥3) = 0.8600

C. P(X=7) = 0.0900, P(X≥3) = 0.8900

D. P(X=7) = 0.1100, P(X≥3) = 0.8800

In [44]:
q1 = poisson.pmf(k=7,mu=5)
print(f'Probability of exactly 7 messags in a minute {q1}')

q2 = 1-poisson.cdf(k=2, mu=5)
print(f'Probability of more than 3 messages in a minute {q2}')

Probability of exactly 7 messags in a minute 0.10444486295705395
Probability of more than 3 messages in a minute 0.8753479805169189


A food delivery app records the time taken for deliveries in minutes:

50% of the orders are delivered in 25 minutes.

30% take 35 minutes.

20% take 45 minutes.

Question 1: What is the expected delivery time?

Question 2: If a customer places 3 independent orders, what is the expected total delivery time?

Options:

A. Expected time = 30 minutes, Total time = 90 minutes

B. Expected time = 32 minutes, Total time = 96 minutes

C. Expected time = 35 minutes, Total time = 105 minutes

D. Expected time = 40 minutes, Total time = 120 minutes

In [47]:
q1 = np.sum([0.5*25,0.3*35,0.2*45])
print(f'Expected delivery time is {q1}')

q2 = q1*3 # number of orders * expected delivery time (q1)
print(f'Total expected delivery time for 3 orders is {q2}')

Expected delivery time is 32.0
Total expected delivery time for 3 orders is 96.0


A job seeker applies to multiple companies via LinkedIn. The probability of receiving a positive response from any company is 15% (0.15).

Question 1: What is the probability that the first positive response comes on the 5th application?

Question 2: What is the expected number of applications the user will need to send before getting a job offer?

In [50]:
q1 = geom.pmf(k=5,p=0.15)
print(f'Probability that the first positive response comes on 5th application is {q1}')

q2 = 1/0.15
print(f'Expected number of applications to get first positive response is {q2}')

Probability that the first positive response comes on 5th application is 0.07830093749999999
Expected number of applications to get first positive response is 6.666666666666667


A Netflix user experiences buffering intervals that follow an Exponential distribution with rate λ= 
15
1
​
  (mean time 15 minutes).

Question 1: What is the probability that the next buffering occurs within 10 minutes?

Question 2: What is the expected time until the next buffering, and what is the variance?

In [52]:
mean = 15
q1 = expon.cdf(10,scale=mean)
print(f'Probability of next buffering less than 10 minutes is {q1}')

q2 = mean
print(f'Expected time until next buffering is {q2}')

q3 = mean**2
print(f'Variance is {q3}')

Probability of next buffering less than 10 minutes is 0.486582880967408
Expected time until next buffering is 15
Variance is 225


In [ ]:
n = len(weights)
    sample_mean = mean(weights)
    sample_std = stdev(weights)


    # Z-scores for common confidence levels
    z_table = {0.90: 1.645, 0.95: 1.96, 0.99: 2.576}
    z = z_table.get(confidence_level, 1.96)  # default to 95% if unknown


    se = sample_std / math.sqrt(n)
    moe = z * se


    lower = round(sample_mean - moe, 2)
    upper = round(sample_mean + moe, 2)


    print(f"Point Estimate (mean): {round(sample_mean, 2)}g")
    print(f"Margin of Error: {round(moe, 2)}g")
    print(f"Confidence Interval: ({lower}g, {upper}g)")

In [60]:
from scipy.stats import poisson
poisson.cdf(k=1,mu=3/20)

np.float64(0.9898141728888165)